### PR TITLE
Add Helix XHarness SDK prop to allow injecting of XHarness .nupkg

### PR DESCRIPTION
**Pull request by premun**
This will allow testing with custom built XHarness (mainly for CI validation in `dotnet/xharness`)

### CODE REVIEW

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.props [hunk 1]
```diff
@@ -1,11 +1,12 @@
 <Project>
   <PropertyGroup>
     <_XHarnessPackageName>Microsoft.DotNet.XHarness.CLI</_XHarnessPackageName>
-    <_XHarnessRunnerPackageFeed Condition="'$(_XHarnessRunnerPackageFeed)'==''">https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json</_XHarnessRunnerPackageFeed>
     <!-- MicrosoftDotNetXHarnessCLIVersion comes from eng\Versions.props -->
     <_XHarnessPackageVersion Condition="'$(_XHarnessPackageVersion)' == ''">$(MicrosoftDotNetXHarnessCLIVersion)</_XHarnessPackageVersion>
     <_HelixMonoQueueTargets>$(_HelixMonoQueueTargets);$(MSBuildThisFileDirectory)XHarnessRunner.targets</_HelixMonoQueueTargets>
 
+    <XHarnessPackageSource Condition="'$(XHarnessPackageSource)'==''">https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json</XHarnessPackageSource>
+    <XHarnessPackageSource Condition="'$(XHarnessNupkgPath)'!=''">$HELIX_WORKITEM_ROOT</XHarnessPackageSource>
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.props:9]: Needs to work on Windows as well.

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets [hunk 2]
```diff
@@ -70,9 +70,16 @@
   </Target>
 
   <Target Name="CreateiOSWorkItems"
-        Condition=" '@(XHarnessAppFolderToTest)' != '' "
-        BeforeTargets="CoreTest">
-    <CreateXHarnessiOSWorkItems AppFolders="@(XHarnessAppFolderToTest)" IsPosixShell="$(IsPosixShell)" WorkItemTimeout="$(XHarnessWorkItemTimeout)" XcodeVersion="$(XHarnessXcodeVersion)">
+          Condition=" '@(XHarnessAppFolderToTest)' != '' "
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets:73]: May as well do this for both, I'm definitely going to want it for Android testing in Xharness repo.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXHarnessiOSWorkItems.cs [hunk 3]
```diff
@@ -129,6 +140,17 @@ private async Task<string> CreateZipArchiveOfFolder(string folderToZip)
             using FileStream payloadScriptStream = GetPayloadScriptStream();
             await payloadScriptStream.CopyToAsync(zipEntryWriter.BaseStream);
 
+            // Add the .nupkg
+            if (!string.IsNullOrEmpty(XHarnessNupkgPath))
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXHarnessiOSWorkItems.cs:144]: Doing this work here doesn't make much sense (The called has no real expectation of an API whose name is "create a zip of this folder" to add things not in the folder to that zip).  Should be added as a standalone correlation payload, ideally in a sub directory.

Comment (premun) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXHarnessiOSWorkItems.cs:144]: Should I also move the payload script that is being added above to a correlation payload?

**Comment by premun** [comment]
@MattGal thanks for the feedback, those were really good points. I moved the code out of the `.cs` files and started adding the .nupkg into a CorrelationPayload.
I am not adding it into a directory because it would have the same name as the file so I didn't see value in that.
This is the real command that was called:
```
dotnet tool install --tool-path /tmp/helix/working/B83409E4/p/xharness-cli --version 1.0.0-dev --add-source /tmp/helix/working/B83409E4/w Microsoft.DotNet.XHarness.CLI
```
[[full logs]](https://helixre107v0xdeko0k025g8.blob.core.windows.net/xharness-master-d199a45baa7d4db88c/xharness-System.Numerics.Vectors.Tests/console.67190018.log?sv=2019-02-02&se=2020-06-23T09%3A49%3A07Z&sr=c&sp=rl&sig=IDdxNTQVhxa1TWyHU1C0rmpLvr73shbi9Ou6VBvuJE4%3D)

File: src/Microsoft.DotNet.Helix/Sdk/CreateXHarnessiOSWorkItems.cs (hunk 3 above)
Comment (premun) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXHarnessiOSWorkItems.cs]: Gotcha

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets [hunk 4]
```diff
@@ -11,31 +11,45 @@
...
+
+    <PropertyGroup Condition=" '$(XHarnessNupkgPath)' != '' AND $(IsPosixShell) ">
+      <XHarnessPackageSource>$HELIX_CORRELATION_ARCHIVE_ROOT</XHarnessPackageSource>
+    </PropertyGroup>
+    <PropertyGroup Condition=" '$(XHarnessNupkgPath)' != '' AND !$(IsPosixShell) ">
+      <XHarnessPackageSource>%HELIX_CORRELATION_ARCHIVE_ROOT%</XHarnessPackageSource>
+    </PropertyGroup>
+
+    <ItemGroup Condition=" '$(XHarnessNupkgPath)' != '' ">
+      <HelixCorrelationPayload Include="$(XHarnessNupkgPath)" />
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets:28]: This a bit funky.  I guess it works but what you end up doing here is unzipping the nupkg to HELIX_CORRELATION_PAYLOAD... raw paths from inside the nupkg all over %HELIX_CORRELATION_PAYLOAD% then never using it.  

While it's slightly more work and no fewer bytes going over the pipe, I really prefer we zip up the nupkg and have it as a payload

**Comment by premun** [comment]
I will check this in to get the bits to runtime and fix it in a follow-up PR

File: src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets (hunk 4 above)
Comment (premun) [reply src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets:28]: I tested this and it doesn't do what you're saying. The nupkg gets zipped by Helix and unpacked as a single .nupkg file inside the correlation payload directory.

I added quite a lot of code to zip the nupkg first and ended up with a zip of a zip of a nupkg which I would have to then extract. The way it's written now is super simple and ends up doing exactly what we want

Comment (MattGal) [reply src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets:28]: I don't think you're following me here but you already merged so let's discuss some other time.

**Pull request merged**